[beta] Added documentation for the useMemo() API reference. (#4928)
reactjs/reactjs.org: beta/src/pages/apis/react/useMemo.md

### Skip expensive recalculations {/*skip-expensive-recalculations*/}

Call `useMemo` at the top level of your component to declare one or more memoized variables.

```js {5}
import {useMemo} from 'react';

export default function App(){

const visibleTodos = useMemo(() => getVisibleTodos(todos, tab), [todos, tab]);

//...
```

Wrap the function that has the expensive calculations with `useMemo`, so that when the dependencies are unchanged, the calculation of the expensive function is skipped.

In the example below, the function `getVisibleTodos` runs only when either one of the dependencies, i.e., `todos` or `tab` change, and the memoized value gets stored in `visibleTodos`.

``` js {4}
//...
const [todos, setTodos] = useState(createInitialTodos);
const [tab, setTab] = useState("all");
const visibleTodos = useMemo(() => getVisibleTodos(todos, tab), [todos, tab]);

function getVisibleTodos(todos, tab) {
  console.log("Running getVisibleTodos for " + todos.length + " todos");
  return todos.filter((t) => {
    if (tab === "completed") {
      return t.completed;
    }
    return true;
  });
}

//...

```

### Skip re-render of components {/*skip-re-render-of-components*/}

> Often you would want to skip re-render of components when not necessary.

Author: We can add a note, stating the recommended way to use `useMemo`.

                type="checkbox"
                checked={item.completed}
                onChange={(e) => {
                  toggleTodo(item.id);
                }}
              />
              {item.text}
            </label>
          </li>
        ))}
      </ul>
    );
  }
  TodoList = memo(TodoList);

```

</Sandpack>

<Solution />

</Recipes>

<DeepDive title="Pros and Cons of using `useMemo`">
Pros

`useMemo` only recomputes the memoized value when one of the dependencies has changed. So, re-computing an expensive function on every render is unnecessary.

Cons

Every time a component is rendered:
* an extra call is made to `useMemo`, thus creating performance overhead
* memory gets allocated to hold the memoized variable value, thus creating memory overhead

</DeepDive>

## Reference {/*reference*/}

### `useMemo(() => computeExpensiveValue(a, b), [a, b])` {/*usememo*/}

Call `useMemo` at the top level of your component to declare one or more memoized variables.

```js
import {useMemo} from 'react';
  // ...
```

#### Parameters {/*parameters*/}

* `computeExpensiveValue`: The parameter is the expensive computational function that requires memoization.

> * `[a,b]`: This value represents the dependency on which the recompute of the `computeExpensiveValue` rests. There can be an array of dependencies or a single value.

Reviewer: Should this be `deps`? `[a,b]` is confusing because there's not always two deps.

#### Returns {/*returns*/}

> `useMemo` returns a value or object depending on the computed value:

Reviewer: ```suggestion
`useMemo` returns a memoized value for the result of `computeExpensiveValue`.
```
